### ISSUE TITLE
Transformer论文精读.md中，Encoder和Decoder中有冗余的LayerNorm层

### ISSUE THREAD

**Issue by Evelyn-ELF** [Hoper-J/AI-Guide-and-Demos-zh_CN#20]
如下图，Encoder和Decoder的每一层里面已经做过“残差连接+层归一化”（先残差连接，后层归一化），但最后又做了一遍层归一化，最后的层归一化是没必要的吧

<img width="1440" height="716" alt="Image" src="https://github.com/user-attachments/assets/49c042e7-5111-4d44-af94-73a727749de6" />

<img width="1282" height="1282" alt="Image" src="https://github.com/user-attachments/assets/4f2883b1-3306-44c7-95ed-488cf80236c3" />

<img width="1288" height="1354" alt="Image" src="https://github.com/user-attachments/assets/0afb4dd9-972c-4239-a8f4-36338fcec624" />

**Comment by Hoper-J** [comment]
感谢同学指出这个容易让人疑惑的点，你是对的。这个问题早期在社区有一些讨论：
- [Successive Layer Normalization in nn.Transformer #24930](https://github.com/pytorch/pytorch/issues/24930)
- [Default nn.Transformer applies layer normalization twice at last #50086](https://github.com/pytorch/pytorch/issues/50086)
- [I think your transformer implementation is flawed #55270](https://github.com/pytorch/pytorch/issues/55270)
- [Two consecutive nn.LayerNorm are used in transformer model when norm_first is False #74092](https://github.com/pytorch/pytorch/issues/74092)

同时也存在 PR 对其进行修复：
[[WIP] fix: no duplicate final LayerNorm in Transformer #74237](https://github.com/pytorch/pytorch/pull/74237)
- 但最终考虑到向后兼容，作者没有接受：
  [<img width="917" height="274" alt="Image" src="https://github.com/user-attachments/assets/694c4876-b35c-4f02-82d2-07ede83de11b" />](https://github.com/pytorch/pytorch/issues/74092#issuecomment-1068308466)

因为文章代码部分的实现最初主要对齐的是 PyTorch 版本的 Transformer 架构，所以问题同样继承到了当前项目。我会抽空修正这一行为，并在最后与 nn.Transformer() 对比时引用当前 issue。

---

**拓展**：如果想了解 PyTorch 版本的 Transformer 的具体结构，可以运行下面的代码：
```python
import torch.nn as nn

# 使用 Transformer base 参数
d_model = 512      # 嵌入维度
N = 6              # 编码器和解码器的层数
h = 8              # 多头注意力的头数
d_ff = 2048        # 前馈神经网络的隐藏层维度
dropout = 0.1      # Dropout 概率

model = nn.Transformer(
    d_model=d_model,
    nhead=h,
    num_encoder_layers=N,
    num_decoder_layers=N,
    dim_feedforward=d_ff,
    dropout=dropout,
    batch_first=True
)

print(model)
```

**输出**：

```
Transformer(
  (encoder): TransformerEncoder(
    (layers): ModuleList(
      (0-5): 6 x TransformerEncoderLayer(
        (self_attn): MultiheadAttention(
          (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
        )
        (linear1): Linear(in_features=512, out_features=2048, bias=True)
        (dropout): Dropout(p=0.1, inplace=False)
        (linear2): Linear(in_features=2048, out_features=512, bias=True)
        (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (dropout1): Dropout(p=0.1, inplace=False)
        (dropout2): Dropout(p=0.1, inplace=False)
      )
    )
    (norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
  )
  (decoder): TransformerDecoder(
    (layers): ModuleList(
      (0-5): 6 x TransformerDecoderLayer(
        (self_attn): MultiheadAttention(
          (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
        )
        (multihead_attn): MultiheadAttention(
          (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
        )
        (linear1): Linear(in_features=512, out_features=2048, bias=True)
        (dropout): Dropout(p=0.1, inplace=False)
        (linear2): Linear(in_features=2048, out_features=512, bias=True)
        (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (norm3): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
        (dropout1): Dropout(p=0.1, inplace=False)
        (dropout2): Dropout(p=0.1, inplace=False)
        (dropout3): Dropout(p=0.1, inplace=False)
      )
    )
    (norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
  )
)
```

可以看到默认情况下，`encoder` 和 `decoder` 最后都额外存在一层 `norm`。

不过需要注意的是，正常实例化 TransformerEncoder(...)  和 TransformerDecoder(...) 默认不会有这一层 `norm`。

```python
import torch.nn as nn

# 使用 Transformer base 参数
d_model = 512      # 嵌入维度
N = 6              # 编码器和解码器的层数
h = 8              # 多头注意力的头数
d_ff = 2048        # 前馈神经网络的隐藏层维度
dropout = 0.1      # Dropout 概率

encoder_layer = nn.TransformerEncoderLayer(
    d_model=d_model,
    nhead=h,
    dim_feedforward=d_ff,
    dropout=dropout,
    batch_first=True,
)
decoder_layer = nn.TransformerDecoderLayer(
    d_model=d_model,
    nhead=h,
    dim_feedforward=d_ff,
    dropout=dropout,
    batch_first=True,
)

encoder = nn.TransformerEncoder(encoder_layer, num_layers=N)
decoder = nn.TransformerDecoder(decoder_layer, num_layers=N)

print(encoder)
print(decoder)
print("encoder.norm is None ->", encoder.norm is None)
print("decoder.norm is None ->", decoder.norm is None)
```

**输出**：
```
TransformerEncoder(
  (layers): ModuleList(
    (0-5): 6 x TransformerEncoderLayer(
      (self_attn): MultiheadAttention(
        (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
      )
      (linear1): Linear(in_features=512, out_features=2048, bias=True)
      (dropout): Dropout(p=0.1, inplace=False)
      (linear2): Linear(in_features=2048, out_features=512, bias=True)
      (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      (dropout1): Dropout(p=0.1, inplace=False)
      (dropout2): Dropout(p=0.1, inplace=False)
    )
  )
)
TransformerDecoder(
  (layers): ModuleList(
    (0-5): 6 x TransformerDecoderLayer(
      (self_attn): MultiheadAttention(
        (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
      )
      (multihead_attn): MultiheadAttention(
        (out_proj): NonDynamicallyQuantizableLinear(in_features=512, out_features=512, bias=True)
      )
      (linear1): Linear(in_features=512, out_features=2048, bias=True)
      (dropout): Dropout(p=0.1, inplace=False)
      (linear2): Linear(in_features=2048, out_features=512, bias=True)
      (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      (norm2): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      (norm3): LayerNorm((512,), eps=1e-05, elementwise_affine=True)
      (dropout1): Dropout(p=0.1, inplace=False)
      (dropout2): Dropout(p=0.1, inplace=False)
      (dropout3): Dropout(p=0.1, inplace=False)
    )
  )
)
encoder.norm is None -> True
decoder.norm is None -> True
```

因为 Transformer(...) 在内部会创建 [encoder_norm](https://github.com/pytorch/pytorch/blob/603ec7f00e46290009a7e1ef01fa69bb4c9f34e2/torch/nn/modules/transformer.py#L137) 和 [decoder_norm](https://github.com/pytorch/pytorch/blob/603ec7f00e46290009a7e1ef01fa69bb4c9f34e2/torch/nn/modules/transformer.py#L162)，并分别传给了 TransformerEncoder(..., encoder_norm) 和 TransformerDecoder(..., decoder_norm)，而只要指定 `norm` 就会触发 [forward()](https://github.com/pytorch/pytorch/blob/603ec7f00e46290009a7e1ef01fa69bb4c9f34e2/torch/nn/modules/transformer.py#L548) 中的代码：
```python
if self.norm is not None:
    output = self.norm(output)
```
